### ISSUE TITLE
False positive: no-value-for-parameter when called decorator is itself decorated

### ISSUE THREAD

**Issue by pylint-bot** [PyCQA/pylint#207]
Originally reported by: **BitBucket: [nathanielmanista](http://bitbucket.org/nathanielmanista)**

---

On line 29 of the attached reproduction example, it's nice that pylint tries to determine that the expression in the decorator is a legitimate function call, but it should abort its analysis when it looks up the function being called and sees that that function is itself decorated (and thus its overall type is probably not known or at least very hard to reason about).

This is happening in pylint 1.0.0 and pylint 1.1.0.

The real-world source of this report is in the Melange project (http://code.google.com/p/soc) at a place where we use the App Engine Datastore's NDB API's ndb.transactional decorator with an argument of "xg=True" - ndb.transactional is a decorator that is itself decorated in a way that seems beyond pylint's reasoning. This defect likely affects all users of the App Engine Datastore's NDB API.

This issue was previously reported over at https://bitbucket.org/logilab/pylint.org/issue/6.

---
- Bitbucket: https://bitbucket.org/logilab/pylint/issue/207

**Comment by kxepal** [comment]
For the reference: this bug reproduced with Hypothesis (with quite all modern versions):
```
from hypothesis import strategies as st
from hypothesis import find


def things():
    @st.composite
    def d(draw):
        return draw(st.text(min_size=1))
    return d()


def list_of_things():
    return st.lists(things(), min_size=1)

assert find(list_of_things(), lambda _: True) == ['0']
```

pylint 1.5.4 gives assertion on the line 9 `return d()`:
```
E:  9,11: No value for argument 'draw' in function call (no-value-for-parameter)
```

The `composite` itself is a decorator that [rewrites](https://github.com/DRMacIver/hypothesis/blob/2.0.0/src/hypothesis/strategies.py#L832-L840) function signature in own fashion.

**Comment by PCManticore** [comment]
Not sure that ``composite`` is something that pylint can understand out of the box. This feels more suited for an astroid brain tip, similar to what we're already having for other modules.

**Comment by deckar01** [comment]
Here is a minimal example to reproduce the issue:

```py
def helper(fun):
    def wrapper(b):
        return fun('helper', b)
    return wrapper

@helper
def simple(a, b):
    print a, b

simple('test')
```

> E: 10, 0: No value for argument 'b' in function call (no-value-for-parameter)

---

Here is the same example with the decorator replaced with method composition:

```py
def helper(fun):
    def wrapper(b):
        return fun('helper', b)
    return wrapper

def hard(a, b):
    print a, b

simple = helper(hard)

simple('test')
```

> No errors

---

To prove that pylint is not simply skipping the signature checking for method composition:

```py
def helper(fun):
    def wrapper(b):
        return fun('helper', b)
    return wrapper

def hard(a, b):
    print a, b

simple = helper(hard)

simple('helper?', 'test')
```

> E: 11, 0: Too many positional arguments for function call (too-many-function-args)

---

The issue is not that pylint is incapable of inferring the signature of of generated methods, but rather that pylint lacks the instruction to treat decorators as the composed functions they are.

**Comment by deckar01** [comment]
This issue seems to stem from the core AST module leaving it up to the consumer to resolve the resulting signature from the `decorator_list`.

It should be sufficient to detect when a function node has decorators and substitute it with the return value(s?) of the first decorator.

I think it would be generally useful to determine the resulting signature in asteroid. Maybe an attribute like `external_args`?

**Comment by Pierre-Sassoulas** [comment]
This issue became unusable because it's referencing old links that are not accessible anymore, please open another issue if you're facing this problem.
